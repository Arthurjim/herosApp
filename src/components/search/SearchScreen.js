import React, { useMemo } from "react";
import { useNavigate, useLocation } from "react-router-dom";
import queryString from "query-string";
import useForm from "../customHooks/useForm";
import HereoCard from "../heroe/HeroeCard";
import { getHerosByName } from "../selectors/getHerosByName";

const SearchScreen = () => {
    // const [hereos, setHeroes] = useState([]);
    const navigate = useNavigate();
    const location = useLocation();
    const { q = "" } = queryString.parse(location.search);

    const [formValues, handleInputChange] = useForm({ searchText: q });

    const { searchText } = formValues;
    const hereos = useMemo(()=> getHerosByName(q),[q]) 
    
    const hanldeSubmit = (e) => {
        e.preventDefault();
        navigate(`?q=${searchText}`);
    };

    return (
        <>
            <div className="row">
                <div className="col-5">
                    <h4 className="text-uppercase">Found your favorit hero.</h4>
                    <hr />

                    <form onSubmit={hanldeSubmit}>
                        <input
                            type="text"
                            placeholder="Search a hero"
                            onChange={handleInputChange}
                            className="form-control"
                            name="searchText"
                            value={searchText}
                            autoComplete="off"
                        />

                        <button
                            type="submit"
                            className="btn btn-outline-primary mt-2"
                        >
                            Search...
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <div className="animate__animated animate__fadeIn">
                        {q === "" ? (
                            <div className="altert alert-info text-center p-2">
                                Heros...
                            </div>
                        ) : (
                            hereos.length === 0 && (
                                <div className="alert alert-danger">
                                    0 heros found
                                </div>
                            )
                        )}

                        {hereos.length > 0 &&
                            hereos.map((hero) => (
                                <HereoCard key={hero.id} {...hero} />
                            ))}
                    </div>
                </div>
            </div>
        </>
    );
};

export default SearchScreen;
