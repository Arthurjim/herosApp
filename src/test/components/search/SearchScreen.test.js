import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import SearchScreen from "../../../components/search/SearchScreen";

const mockNavigate = jest.fn();

//Mocku es una simulaciÃ³n de algo
jest.mock('react-router-dom',()=>({
    ...jest.requireActual('react-router-dom'),
    useNavigate: ()=>mockNavigate,
}))

describe("Pruebas en <SerchScren />", () => {
    test("Debe mostrarse correctamente", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe("Heros...");
    });

    test("Debe de mostrar a batman y el input cone valor del queryString", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper.find("input").prop("value")).toBe("batman");
    });
    test("Debe de mostrar un error si no se encuentra el hero", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=123124"]}>
                <SearchScreen />
            </MemoryRouter>
        );
        expect(wrapper.find(".alert-danger").text().trim()).toBe('0 heros found')
    });

    test('Debe de llamar el navigate a la nueva pantalla ', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=123124"]}>
                <SearchScreen />
            </MemoryRouter>
        );


        wrapper.find('input').simulate('change', {
            target: {name: 'searchText',value:'batman'}

        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault:()=>{},
        })
        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman');

    });
    
});
