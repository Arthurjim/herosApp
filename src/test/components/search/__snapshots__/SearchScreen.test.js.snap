// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SerchScren /> Debe mostrarse correctamente 1`] = `
<div
  className="row"
>
  <div
    className="col-5"
  >
    <h4
      className="text-uppercase"
    >
      Found your favorit hero.
    </h4>
    <hr />
    <form
      onSubmit={[Function]}
    >
      <input
        autoComplete="off"
        className="form-control"
        name="searchText"
        onChange={[Function]}
        placeholder="Search a hero"
        type="text"
        value=""
      />
      <button
        className="btn btn-outline-primary mt-2"
        type="submit"
      >
        Search...
      </button>
    </form>
  </div>
  <div
    className="col-7"
  >
    <div
      className="animate__animated animate__fadeIn"
    >
      <div
        className="altert alert-info text-center p-2"
      >
        Heros...
      </div>
    </div>
  </div>
</div>
`;
